### DISCUSSION TITLE
BrokenBusinessRules for BusinessListBase

### DISCUSSION THREAD

**Discussion by Chicagoan2016** [MarimerLLC/csla#3438] [Questions] [answered]
I have to 'batch' save a collection (read-write) where the child object(s) has business/validation rules.
When I check for IsValid, is there a way to show broken business rules to the users? Since BusinessListBase doesn't have GetBrokenRules method, do we have to loop through the collection to find 'invalid' child objects?
I am not sure if something like the line below will work
```cs
  var allRules = BusinessRules.GetAllBrokenRules(colllineitems);
```

Kind Regards

**Comment by rockfordlhotka** [comment, marked as answer]
It depends on what you need to show the user.

A simple `foreach` loop through the child objects allows you to call `GetBrokenRules` on each and build a list of all the broken rules.

I see a challenge here, in that the `BrokenRule` type doesn't keep track of the object it pertains to, so you can't tell the user which object corresponds to each broken rule without doing that work yourself. You could solve that in a number of ways if that information is relevant.

For example, define a `MyBrokenRule` class:

```c#
  public class MyBrokenRule
  {
    public MyBrokenRule(MyChild child, Csla.Rules.BrokenRule rule)
    {
      ContainingObject  = child;
      BrokenRule = rule;
    }

    public MyChild ContainingObject { get; set; }
    public Csla.Rules.BrokenRule BrokenRule { get; set; }
  }
```

Then add this to the list class:

```c#
    public List<MyBrokenRule> GetBrokenRules()
    {
      var result = new List<MyBrokenRule>();
      foreach (var item in this)
      {
        foreach (var rule in item.GetBrokenRules())
        {
          result.Add(new MyBrokenRule(item, rule));
        }
      }
      return result;
    }
```

Then your calling code can display the broken rules for all children, with info about each child like this:

```c#
var rules = list.GetBrokenRules();
foreach (var rule in rules)
  Console.WriteLine($"{rule.ContainingObject}, {rule.BrokenRule}");
```

**Comment by Chicagoan2016** [comment]
@rockfordlhotka Thank you Sir.
It worked perfectly. I put MyBrokenRule class inside my collection (read-write list).
Is there a way we could make it 'generic' so it would work with all (my custom) collections?
Kind Regards
